Fix GroupId with aliased grouping key columns

aditi-pandit opened the pull request:
GROUPING SETS can be specified with grouping keys that are alias columns of the same input column. This is typically used to compute multiple mixed aggregations on the same key.

e.g.
`select COUNT(orderkey), count(distinct orderkey) from orders;` 
or

`SELECT lna, lnb, SUM(quantity) FROM (SELECT linenumber lna, linenumber lnb, CAST(quantity AS BIGINT) quantity FROM lineitem) GROUP BY GROUPING SETS ((lna, lnb), (lna), (lnb), ())
`

The Velox operator always assumes that : 

- An input column maps to a single output grouping key. 

The Prestissimo code assumes that
- A GroupingSet is specified using input column names. Presto co-ordinator uses output column names for Grouping Sets allowing for multiple uses of the same input column. 

Velox/Presto behavior lead to incorrect results. In each grouping set the column or its alias column values all appear in the same output column leading to wrong computations.

Fixes https://github.com/prestodb/presto/issues/20910 and https://github.com/prestodb/presto/issues/20917
Prestissimo side fix https://github.com/prestodb/presto/pull/20964 is also needed.

mbasmanova commented on velox/exec/tests/utils/PlanBuilder.h:
```diff
@@ -477,12 +477,18 @@ class PlanBuilder {
+  /// The extraGroupingKeys specify the additional projections of output
+  /// key columns to the input columns. This allows grouping sets with
+  /// multiple groups corresponding to the same input key.
+  /// Note: Grouping sets are specified using output key columns.
```
This is very confusing. 

My understanding is GroupIdNode effectively includes an optional projection that may duplicate and/or rename some input columns.

aditi-pandit replied:
Thanks. I've refactored the API.

mbasmanova reviewed (commented):
@aditi-pandit Aditi, thank you for investigating this issue and coming up with a fix. This is very confusing, so let's figure out how to make this clear for future readers. Let's update comments for GroupIdNode and documentation in https://facebookincubator.github.io/velox/develop/operators.html#groupidnode. I feel that verbal explanation won't be sufficient and wondering if we could add a few examples of GroupNodeId configurations along with sample input and output.

aditi-pandit commented:
> @aditi-pandit Aditi, thank you for investigating this issue and coming up with a fix. This is very confusing, so let's figure out how to make this clear for future readers. Let's update comments for GroupIdNode and documentation in https://facebookincubator.github.io/velox/develop/operators.html#groupidnode. I feel that verbal explanation won't be sufficient and wondering if we could add a few examples of GroupNodeId configurations along with sample input and output.

@mbasmanova : Thanks Masha.

Couple of points:

- The GroupID operator API remains the same with this change. I clarified in operators.rst that Grouping keys are specified with output names.
- What was modified was the GroupId API in PlanBuilder since it didn't allow any way to specify an alias column. It built grouping keys from groupingSets. I have now changed that API to specify grouping keys separately so alias columns can be added there. I have also updated its documentation.

In terms of fixes there are 2 changes:
i) The Velox operator didn't take into account the case that a single input column could be mapped to 2. So I changed std::unordered_map<std::string, column_index_t>
      inputToOutputGroupingKeyMapping; to an output to input mapping of grouping key columns. This was used further in the operator.
ii) The code on the Prestissimo side had mapped output names to input names when filling GroupingSets. I changed that code to send the output names. That is the fix on the Prestissimo side.

Hope that clarifies the fixes. Please let me know if you have further questions.

aditi-pandit commented:
@mbasmanova : Changes made:

- Add SQL example in operators.rst
- Add new field groupingSetsColumns_ for using output column names in groupingSet. This will deprecate the original groupingSets_  variable that uses FieldAccessTypeExprPtr. I will remove this code after changing Prestissimo code to use the new field.
- Change PlanBuilder to pick up ordering of grouping keys from GroupingKeys instead of the order of keys in groupingSets.

mbasmanova commented on velox/core/PlanNode.h:
```diff
@@ -785,6 +785,9 @@ class GroupIdNode : public PlanNode {
     folly::dynamic serialize() const;
   };
 
+  /// Note : This GroupIdNode constructor is deprecated. It uses deprecated
```
Would you wrap this in VELOX_ENABLE_BACKWARD_COMPATIBILITY macro?

mbasmanova commented on velox/core/PlanNode.h:
```diff
@@ -845,7 +871,21 @@ class GroupIdNode : public PlanNode {
+  // use for the groupingSet. Note : The GroupIdNode should have only one of
+  // groupingSetsWithColumns_ or groupingSets_.
+  const std::vector<std::vector<std::string>> groupingSetsWithColumns_;
+  // Note: Deprecated by groupingSetsWithColumns_; This will be removed after
```
VELOX_ENABLE_BACKWARD_COMPATIBILITY

mbasmanova commented on velox/core/PlanNode.h:
```diff
@@ -816,6 +838,10 @@ class GroupIdNode : public PlanNode {
     return groupingSets_;
   }
 
+  const std::vector<std::vector<std::string>>& groupingSetsWithColumns() const {
```
Is groupingSets() method used in Prestissimo? If not, let's reuse that name.

aditi-pandit replied:
Sure.

mbasmanova commented on velox/docs/develop/operators.rst:
```diff
@@ -232,13 +232,54 @@ followed by the group ID column. The type of group ID column is BIGINT.
 
+To illustrate why GroupingSets should use output column names lets examine the following SQL query:
+.. code-block:: sql
+    select COUNT(orderkey), count(distinct orderkey) from orders as t(orderkey);
```
Would you generate the docs locally and verify formatting. I think you may need an empty line.

mbasmanova commented on velox/docs/develop/operators.rst:
```diff
@@ -232,13 +232,54 @@ followed by the group ID column. The type of group ID column is BIGINT.
+are related and can be computed in a staged way using multiple grouping sets using the same input column as grouping key.
+Presto generates the following plan for the query:
+
+     - Output[PlanNodeId 11][_col0, _col1] => [count:bigint, count_3:bigint]
```
This is not very readable. It would be nice to explain how Presto uses grouping sets to compute this query and provide a sketch of a query plan.

mbasmanova replied:
> select COUNT(orderkey), count(distinct orderkey) from orders

I'm just trying to understand how Presto plans this query.

It first uses GroupId node to duplicate every record and assign one copy to group 0 and another copy to group 1.

Let's say we have 100 rows with orderkeys 1...100:

```
1
2
...
100
```

After GroupIdNode we get 200 rows

```
k1 k2 gId
1  null 0
2  null 0
...
100 null 0
null 1 1
null 2 1
...
null 100 1 
```

Then, Presto plans partial + final aggregation with grouping key k2, gid and count(k1) aggregate. This produces 101 rows.

```
k2 gid c
null 0 100
1 1 null
2 1 null
...
100 1 null
```

Then, Presto plans another aggregation with no grouping keys and two aggregates: count(k2) and arbitrary(c). This produces 1 row.

```
count(k2) arbitrary(c)
100 100
```

Both count and arbitrary functions ignore null inputs, hence, the results are correct.

I imagine that the same result could be achieved without a GroupId node. First, group by orderkey and count(). Then sum(count) + count.

I wonder why Presto uses GroupId node in this case. Is there another scenario which illustrates that it is necessary? Am I missing anything?

aditi-pandit replied:
For this specific query, the plan was generated when optimize-mixed-distinct-aggregations = true. While this might seem specific, it was a better example to illustrate than some other SQL.

In the tests there was also this example which used GroupId as directed.
SELECT lna, lnb, SUM(quantity) FROM (SELECT linenumber lna, linenumber lnb, CAST(quantity AS BIGINT) quantity FROM lineitem) GROUP BY GROUPING SETS ((lna, lnb), (lna), (lnb), ()) 
Don't know if this example is better.
wdyt ?

aditi-pandit replied:
Yes, the GROUPING SETS example seemed contrived.

I see what kind of explanation you are suggesting. Will do an update.

mbasmanova reviewed (commented):
@aditi-pandit Thank you for iterating on this. Some questions. I'll take a closer look a bit later.

mbasmanova commented on velox/core/PlanNode.cpp:
```diff
@@ -401,13 +401,53 @@ GroupIdNode::GroupIdNode(
+      groupingKeyInfos_(std::move(groupingKeyInfos)),
+      aggregationInputs_(std::move(aggregationInputs)),
+      groupIdName_(std::move(groupIdName)) {
+  VELOX_CHECK_GE(
```
USER check

mbasmanova commented on velox/exec/tests/AggregationTest.cpp:
```diff
@@ -1500,14 +1500,15 @@ TEST_F(AggregationTest, groupingSets) {
+  plan =
+      PlanBuilder()
+          .values({data})
+          .groupId({"k1", "k2"}, {{"k1", "k2"}, {"k1"}, {"k2"}, {}}, {"a", "b"})
```
These test cases were meant to map to well-known use cases. Let's not change them, but rather add new uses cases that we didn't know and didn't support before.

aditi-pandit replied:
Have changed to a test inspired by a GROUPING SETS based SQL.

mbasmanova commented on velox/exec/tests/AggregationTest.cpp:
```diff
@@ -1591,6 +1592,38 @@ TEST_F(AggregationTest, groupingSetsOutput) {
+  auto data = makeRowVector(
+      {"k1", "a", "b"},
+      {
+          makeFlatVector<int64_t>(size, [](auto row) { return row; }),
```
When making short vectors (5 rows), it is more readable to list out the values:

```
makeFlatVector<int64_t>({0, 1, 2, 3, 4});
makeFlatVector<int64_t>({0, 1, 2, 3, 4});
makeFlatVector<std::string>({"", "x", "xx", "xxx", "xxxx});
```

mbasmanova commented on velox/exec/tests/AggregationTest.cpp:
```diff
@@ -1591,6 +1592,38 @@ TEST_F(AggregationTest, groupingSetsOutput) {
+              size,
+              [](auto row) {
+                auto str = std::string(row % 12, 'x');
+                return StringView(str);
```
It is not safe to return StringView over temporary memory that goes away while StringView is still alive.

aditi-pandit replied:
This was copied from functions above. I'll change those in another PR.

aditi-pandit replied:
https://github.com/facebookincubator/velox/pull/6905

mbasmanova commented on velox/exec/GroupId.cpp:
```diff
@@ -29,27 +29,39 @@ GroupId::GroupId(
+          : groupIdNode->groupingSetsWithColumns()[i][j];
+
+      auto outputChannel = outputType_->getChildIdx(outputKeyName);
+      VELOX_CHECK(
```
VELOX_USER_CHECK_NE

aditi-pandit commented:
@mbasmanova : Have fixed the docs and make the backward compatibility code simpler. PTAL.

mbasmanova commented on velox/core/PlanNode.h:
```diff
@@ -803,6 +803,25 @@ class GroupIdNode : public PlanNode {
       std::string groupIdName,
       PlanNodePtr source);
 
+  /// @param id Plan node ID.
```
Can we still wrap the old constructor in a macro? The implementation would need to be moved to the .h file for that old constructor.

aditi-pandit replied:
Moving the constructor code over involves moving functions getGroupIdOutputType and getGroupingSets also which seemed cumbersome.
I've added a note about the function deprecation.

mbasmanova commented on velox/exec/tests/AggregationTest.cpp:
```diff
@@ -1591,6 +1592,32 @@ TEST_F(AggregationTest, groupingSetsOutput) {
+  auto plan = PlanBuilder()
+                  .values({data})
+                  .groupId(
+                      {"o_key", "o_key as o_key_1"},
```
I still find this use case confusing. It is just not clear why would anything use a plan like this. I wish we could identify a compelling use case.

aditi-pandit replied:
I felt the query documented with its Presto plan was a better example. But I wasn't able to translate the IF expressions to equivalent CASE expressions. `case group_id when 1 then 0 else null end` could not compile since null is UNKNOWN type and CASE required all when/then expressions to evaluate to the same type.

mbasmanova replied:
try "null::bigint"

aditi-pandit replied:
I will send out a follow up PR.

mbasmanova commented on velox/docs/develop/operators.rst:
```diff
@@ -232,13 +232,77 @@ followed by the group ID column. The type of group ID column is BIGINT.
+
+Lets say the orders table has 10 rows as follows:
+
+.. code-block::
```
Would you share a screenshot of the docs? Curious how they look? Would it be better to format this as a table?

aditi-pandit replied:
![image](https://github.com/facebookincubator/velox/assets/87670623/c7389eee-c6e1-436f-bbe6-3a41cd5078ce)

aditi-pandit replied:
This seemed okay to me.

mbasmanova commented on velox/docs/develop/operators.rst:
```diff
@@ -232,13 +232,77 @@ followed by the group ID column. The type of group ID column is BIGINT.
    * - groupIdName
      - The name for the group-id column that identifies the grouping set. Zero-based integer corresponding to the position of the grouping set in the 'groupingSets' list.
 
+To illustrate why GroupingSets should use output column names lets examine the following SQL query:
```
It would be nice to first explain that GroupId node can be used to compute cubes and rollups. Then elaborate that sometimes we want to create separate groping sets using same grouping keys.

mbasmanova commented on velox/docs/develop/operators.rst:
```diff
@@ -232,13 +232,77 @@ followed by the group ID column. The type of group ID column is BIGINT.
+
+  select COUNT(orderkey), count(distinct orderkey) from orders;
+
+In this query the user wants to compute aggregates on the same key, though with
```
> aggregates on the same key

This query doesn't have grouping keys. The aggregations are global aggregations computed over the same input column.

mbasmanova commented on velox/docs/develop/operators.rst:
```diff
@@ -232,13 +232,77 @@ followed by the group ID column. The type of group ID column is BIGINT.
+
+.. code-block:: sql
+
+  select COUNT(orderkey), count(distinct orderkey) from orders;
```
capitalize SQL keywords

```
SELECT count(orderkey), count(DISTINCT orderkey) FROM orders;
```

mbasmanova commented on velox/docs/develop/operators.rst:
```diff
@@ -232,13 +232,77 @@ followed by the group ID column. The type of group ID column is BIGINT.
+
+In this query the user wants to compute aggregates on the same key, though with
+and without the DISTINCT clause. With a particular optimization strategy
+optimize.mixed-distinct-aggregations, Presto uses GroupIdNode to compute these.
```
Can we link here to Presto's documentation about this property?

aditi-pandit replied:
I just found https://prestodb.io/docs/current/release/release-0.156.html?highlight=mixed%20distinct%20aggregations. Do you want to link this ?

mbasmanova replied:
I found this post which seems useful to link to: https://www.qubole.com/blog/presto-optimizes-aggregations-over-distinct-values

mbasmanova replied:
Also this seems useful: https://github.com/trinodb/trino/pull/15927

aditi-pandit replied:
Thanks for the link. Have added it.

mbasmanova commented on velox/docs/develop/operators.rst:
```diff
@@ -232,13 +232,77 @@ followed by the group ID column. The type of group ID column is BIGINT.
+and without the DISTINCT clause. With a particular optimization strategy
+optimize.mixed-distinct-aggregations, Presto uses GroupIdNode to compute these.
+
+First, the optimizer duplicates every row assigning one copy to group 0 and another
```
> optimizer duplicates every row

Optimizer creates a plan to duplicate .... Optimizer doesn't duplicate rows.

mbasmanova commented on velox/docs/develop/operators.rst:
```diff
@@ -232,13 +232,77 @@ followed by the group ID column. The type of group ID column is BIGINT.
+groups the orderkey column is aliased as a grouping key for one of the
+grouping sets.
+
+Lets say the orders table has 10 rows as follows:
```
nit: drop "as follows"

mbasmanova commented on velox/docs/develop/operators.rst:
```diff
@@ -232,13 +232,77 @@ followed by the group ID column. The type of group ID column is BIGINT.
+.. code-block::
+
+  orderkey
+     1
```
Let's include some duplicate orderkeys in the example.

mbasmanova reviewed (approved):
@aditi-pandit Thank you for iterating on this PR. Looks great % a few nits.

mbasmanova reviewed (approved):
Looks great. Thanks. 

Any chance you could add a link for "optimize.mixed-distinct-aggregations"?

https://www.qubole.com/blog/presto-optimizes-aggregations-over-distinct-values

mbasmanova commented:
@aditi-pandit I'm seeing e2e test failures:

```
[ERROR] com.facebook.presto.nativeworker.TestPrestoNativeAggregations.testGroupingSets  Time elapsed: 0.327 s  <<< FAILURE!
java.lang.AssertionError: Execution of 'actual' query failed: SELECT orderstatus, orderpriority, count(1), min(orderkey) FROM orders GROUP BY GROUPING SETS ((orderstatus), (orderpriority))
	at org.testng.Assert.fail(Assert.java:98)
	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:178)
	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:106)
	at com.facebook.presto.tests.AbstractTestQueryFramework.assertQuery(AbstractTestQueryFramework.java:152)
	at com.facebook.presto.tests.AbstractTestQueryFramework.assertQuery(AbstractTestQueryFramework.java:147)
	at com.facebook.presto.nativeworker.AbstractTestNativeAggregations.testGroupingSets(AbstractTestNativeAggregations.java:114)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:135)
	at org.testng.internal.invokers.TestInvoker.invokeMethod(TestInvoker.java:673)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethod(TestInvoker.java:220)
	at org.testng.internal.invokers.MethodRunner.runInSequence(MethodRunner.java:50)
	at org.testng.internal.invokers.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:945)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethods(TestInvoker.java:193)
	at org.testng.internal.invokers.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.RuntimeException:  Field not found: orderstatus. Available fields are: orderpriority$gid, orderstatus$gid, expr, orderkey, groupid.
	at com.facebook.presto.tests.AbstractTestingPrestoClient.execute(AbstractTestingPrestoClient.java:124)
	at com.facebook.presto.tests.DistributedQueryRunner.execute(DistributedQueryRunner.java:733)
	at com.facebook.presto.tests.DistributedQueryRunner.execute(DistributedQueryRunner.java:701)
	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:175)
	... 19 more
Caused by: VeloxUserError:  Field not found: orderstatus. Available fields are: orderpriority$gid, orderstatus$gid, expr, orderkey, groupid.
```

aditi-pandit commented:
> @aditi-pandit I'm seeing e2e test failures:
> 
> ```
> [ERROR] com.facebook.presto.nativeworker.TestPrestoNativeAggregations.testGroupingSets  Time elapsed: 0.327 s  <<< FAILURE!
> java.lang.AssertionError: Execution of 'actual' query failed: SELECT orderstatus, orderpriority, count(1), min(orderkey) FROM orders GROUP BY GROUPING SETS ((orderstatus), (orderpriority))
> 	at org.testng.Assert.fail(Assert.java:98)
> 	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:178)
> 	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:106)
> 	at com.facebook.presto.tests.AbstractTestQueryFramework.assertQuery(AbstractTestQueryFramework.java:152)
> 	at com.facebook.presto.tests.AbstractTestQueryFramework.assertQuery(AbstractTestQueryFramework.java:147)
> 	at ...
> Caused by: java.lang.RuntimeException:  Field not found: orderstatus. Available fields are: orderpriority$gid, orderstatus$gid, expr, orderkey, groupid.
> 	at com.facebook.presto.tests.AbstractTestingPrestoClient.execute(AbstractTestingPrestoClient.java:124)
> 	at com.facebook.presto.tests.DistributedQueryRunner.execute(DistributedQueryRunner.java:733)
> 	at com.facebook.presto.tests.DistributedQueryRunner.execute(DistributedQueryRunner.java:701)
> 	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:175)
> 	... 19 more
> Caused by: VeloxUserError:  Field not found: orderstatus. Available fields are: orderpriority$gid, orderstatus$gid, expr, orderkey, groupid.
> ```

@mbasmanova : Taking a look.

aditi-pandit commented:
@mbasmanova : Have fixed the issue. Prestissimo passes groupingFields with input column names so I have to fix this to the output column name in the backward compatible logic of GroupIdNode construction. I had missed that.

Note : The original change needs the follow up https://github.com/prestodb/presto/pull/20964 too
This PR changes code to use input names.

Thanks for your patience.

mbasmanova commented:
@aditi-pandit Aditi, thank you for the quick fix. The changes make sense to me. I'm going to try to merge now.

aditi-pandit commented:
> @aditi-pandit Aditi, thank you for the quick fix. The changes make sense to me. I'm going to try to merge now.

Thanks @mbasmanova. Appreciate the help.

yzhang1991 commented:
@aditi-pandit @mbasmanova Is this merged?
